plugins/ptp_operator/metrics/metrics.go

```diff
@@ -0,0 +1,420 @@
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	ceevent "github.com/redhat-cne/sdk-go/pkg/event"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	ptpNamespace                = "openshift"
+	ptpSubsystem                = "ptp"
+	phc2sysProcessName          = "phc2sys"
+	ptp4lProcessName            = "ptp4l"
+	unLocked                    = "s0"
+	clockStep                   = "s1"
+	locked                      = "s2"
+	maxOffsetThrushHold float64 = 100
+	minOffsetThrushHold float64 = -100
+)
+
+var (
+	// NodeName from the env
+	ptpNodeName = ""
+	eventConfig *common.SCConfiguration
+
+	// OffsetFromMaster metrics for offset from the master
+	OffsetFromMaster = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: ptpNamespace,
+			Subsystem: ptpSubsystem,
+			Name:      "offset_from_master",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+
+	// MaxOffsetFromMaster  metrics for max offset from the master
+	MaxOffsetFromMaster = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: ptpNamespace,
+			Subsystem: ptpSubsystem,
+			Name:      "max_offset_from_master",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+
+	// FrequencyAdjustment metrics to show frequency adjustment
+	FrequencyAdjustment = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: ptpNamespace,
+			Subsystem: ptpSubsystem,
+			Name:      "frequency_adjustment",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+
+	// DelayFromMaster metrics to show delay from the master
+	DelayFromMaster = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: ptpNamespace,
+			Subsystem: ptpSubsystem,
+			Name:      "delay_from_master",
+			Help:      "",
+		}, []string{"process", "node", "iface"})
+)
+
+var registerMetrics sync.Once
+
+// RegisterMetrics ... register metrics for all side car plugins
+func RegisterMetrics(nodeName string) {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(OffsetFromMaster)
+		prometheus.MustRegister(MaxOffsetFromMaster)
+		prometheus.MustRegister(FrequencyAdjustment)
+		prometheus.MustRegister(DelayFromMaster)
+
+		// Including these stats kills performance when Prometheus polls with multiple targets
+		prometheus.Unregister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+		prometheus.Unregister(prometheus.NewGoCollector())
+
+		ptpNodeName = nodeName
+	})
+}
+
+// Stats calculates stats  nolint:unused
+type Stats struct {
+	num                 float64
+	max                 float64
+	min                 float64
+	mean                float64
+	sumSqr              float64
+	sumDiffSqr          float64
+	frequencyAdjustment float64
+	delayFromMaster     float64
+	lastOffset          float64 //nolint:structcheck
+	lastClockState      ceevent.SyncState
+}
+
+func (s *Stats) addValue(val float64) {
+
+	oldMean := s.mean
+
+	if s.max < val {
+		s.max = val
+	}
+	if s.min > val {
+		s.min = val
```

Comment:
If s.min is not initialized then this will always make s.min = 0. I think it makes sense to keep `s.num == 0 || ` here.

Reply:
in Go  variables declared without an initial value are set to their zero values.. so here it will be 0.0

Reply:
Without `s.num == 0 || `, if all values in s are greater than 0, the min is still 0. Is this intended?

Reply:
right ,I  will add it back. we need to capture the first value when its locked as min